pageserver/src/pgdatadir_mapping.rs

```diff
@@ -327,45 +332,49 @@ impl Timeline {
                 )
                 .await?;
 
-            if cmp {
-                high = mid;
-            } else {
+            if !cmp {
+                // We either 1) found commit with timestamp **before** `search_timestamp`;
+                // or 2) we haven't found any commit records at all.
+                // Search with a larger LSN, in case 1) to try to find a more recent commit
+                // (but still **before** target timestamp); and in case 2) to fetch more
+                // SLRU segments for `clog`.
                 low = mid + 1;
+            } else {
+                // We found only more recent commits, search in the older range.
+                high = mid;
             }
         }
+        // If `found_smaller == true`, `low` is the LSN of the first commit record
+        // **before** the `search_timestamp` + 1 (to hit the while loop exit condition).
+        // Substitute 1 to get exactly the commit LSN.
+        let commit_lsn = Lsn((low - 1) * 8);
         match (found_smaller, found_larger) {
             (false, false) => {
                 // This can happen if no commit records have been processed yet, e.g.
                 // just after importing a cluster.
                 Ok(LsnForTimestamp::NoData(max_lsn))
             }
             (true, false) => {
-                // Didn't find any commit timestamps larger than the request
-                Ok(LsnForTimestamp::Future(max_lsn))
+                // Only found a commit with timestamp smaller than the request.
+                // It's still a valid case for branch creation, return it.
+                // And `update_gc_info()` ignores LSN for a `LsnForTimestamp::Future`
+                // case, anyway.
+                Ok(LsnForTimestamp::Future(commit_lsn))
             }
             (false, true) => {
                 // Didn't find any commit timestamps smaller than the request
                 Ok(LsnForTimestamp::Past(max_lsn))
```

Comment:
Not new with this PR, but this actually seems wrong to me. Shouldn't this return `Past(min_lsn)`?

Here's how I think of the return values:

NoData: Not much we can do. (Not sure if returning 'max_lsn' makes sense)

Future(lsn): all commits <= LSN happened before the given timestamp. But because there were no commits after the given timestamp, it seems that the timestamp is in the future. More commits might appear later that are still before the given timestamp. So if you call the function again with the same timestamp later, it might return a different, later LSN.

Past(lsn): All commits within the PITR horizon are newer than the given timestamp. All commits > LSN certainly happened after the given timestamp, but any commits older than the returned LSN might have happened before or after the given timestamp. If you called the function with the same timestamp some time earlier, when the PITR horizon went further back, it might've returned an older LSN.